[MRG] DOC add mixed categorical / continuous example with ColumnTransformer

partmor opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
Fixes #11185
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix?
I include in the example gallery an example for `ColumnTransformer` that uses a dataset with 
categorical / continous features.

#### Any other comments?
I have built the example using a dataset with mixed types stored in a numpy array. I think this is not the most correct way to treat this type of dataset (the array's dtype must be object); but I found it necessary to avoid using pandas.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

partmor committed:
DOC add mixed categorical / continuous example with ColumnTransformer

amueller commented:
Thanks. It would be nice not to require a download, or at least not custom parsing code. But I'm not sure that's realistic. I'm not sure if we want to allow pandas in the examples but we could also just do a ``pd.read_csv`` to an URL, of, say, titanic. That would also give an opportunity to show the typical extraction of the labels from a dataframe.

partmor commented:
@amueller I searched in the `sklearn.datasets` and the only dataset that I saw with mixed types was Boston housing, but the categorical feature was already encoded. So load from URL was the best thing I could think about.

Regarding the use of pandas, I found numpy not very elegant to hold string and float data, furthermore, not using pandas blocks us from demonstrating the true power of using [`sklearn.compose.make_column_transformer`](http://scikit-learn.org/dev/modules/generated/sklearn.compose.make_column_transformer.html#sklearn-compose-make-column-transformer) with column names.

I will procede then to refactor the example using `pd.read_csv` to a *conventional* URL, e.g. titanic.

amueller commented:
I think that's a good idea, and yes, there's no dataset loader for a dataset with mixed features right now. ``fetch_openml`` will provide that "soon".

partmor committed:
DOC change ColumnTransformer example. using pandas now

partmor commented:
I've changed the example. 
Here `ColumnTransformer` takes 2 preprocessing pipelines (for both categorical and numerical data).
I also implemented a custom imputer for categorical data that is used in the mentioned pipeline.

TomDLT commented:
I am surprised `CustomCategoryImputer` is not something we already have.
I might be out of the loop, but if we don't have it, I think we should, maybe in `SimpleImputer` with a new strategy `(strategy='fixed')`, or inside `CategoricalEncoder` (or its re-factored version in #10523).

partmor commented:
@TomDLT I have checked the imputers in `sklearn.impute` (including `SimpleImputer`) and only numerical types are supported (`strategy={'mean', 'median', 'most_frequent}`).

Setting `handle_unknown='ignore'` in `CategoricalEncoder` allows to return all zeros when unseen categories are encountered when transforming, but yet it can't handle `np.nan` (even during `fit`).

jorisvandenbossche commented:
Yes, the inability to handle missing values with categorical data (both in SimpleImputer and CategoricalEncoder )is a know problem, which we hope to fix at least with a basic solution before the release. We were discussing this yesterday, and were also thinking of adding a 'constant'/'fixed' strategy for SimpleImputer. 
But until then, the custom transformer here is fine I think (although it distracts from the actual example),  then we can later simply replace it with the built-in solution.

partmor commented:
@jorisvandenbossche I agree it could distract a bit from the core matter of the example, but I thought it was interesting to include it in order to use a `Pipeline` for the categorical data inside `ColumnTransformer`.

The alternative solution would have been to use pandas' `dropna` on the categorical subset at the beginning of the script, and just apply `CategoricalEncoder`.

Also, I am tempted to extend the example a bit more and include a hyperparameter optimization with grid search for instance, in order to demonstrate how to access the parameters within a `ColumnTransformer.` I'll try to do this later on today.

Any further thoughts?

jorisvandenbossche commented on examples/column_transformer_mixed_types.py:
```diff
@@ -0,0 +1,121 @@
+
+# We will train our classifier with the following features
+num_feats = ['age', 'fare']
+cat_feats = ['embarked', 'sex']
```
I think you can keep 'pclass' as well. It is already integer, but it then shows that that works as well for one-hot encoding.

partmor replied:
Sure, no problem

jorisvandenbossche commented on examples/column_transformer_mixed_types.py:
```diff
@@ -0,0 +1,121 @@
+    CategoricalEncoder('onehot-dense')
+)
+
+preprocessing_pl = ColumnTransformer(
```
should we use `make_column_transformer` for the example?

partmor replied:
I decided to instantiate the object directly because I already saw an example that uses `make_column_transformer` and the latter doesn't have the `remainder` param, which I thought was worth showing.

I wouldn't have any problem using `make_column_transformer` if that's the use we want to encourage.

jorisvandenbossche replied:
There is a PR to add the `remainder` keyword to the `make_column_transformer` (that was an oversight it is not there), so that should at least be fixed soon.

partmor replied:
Ah! great stuff. Will switch to `make_column_transformer` then.

jorisvandenbossche replied:
PR is merged: https://github.com/scikit-learn/scikit-learn/pull/11183

jorisvandenbossche reviewed (commented):
I don't know how 'far' we want to go for this example, but it could also be nice to show that you can do gridsearch on the parameters of the preprocessing steps.

jorisvandenbossche commented:
> The alternative solution would have been to use pandas' dropna on the categorical subset at the beginning of the script, and just apply CategoricalEncoder.

Alternative pandas solution would also be to `fillna('missing')` on the categorical columns (which would be equivalent to your custom transformer). 
Anyhow, I don't think this is too important, as we plan to replace this part with a SimpleImputer in the short term.

> Also, I am tempted to extend the example a bit more and include a hyperparameter optimization with grid search

I commented just the same, so yes, I think that's a good idea :-) 
I did a small GridSearchCV in my example here https://jorisvandenbossche.github.io/blog/2018/05/28/scikit-learn-columntransformer/#Using-our-pipeline-in-a-grid-search

jorisvandenbossche commented:
One other suggestion, certainly if you would make the example longer, is to use the typical sphinx gallery syntax to alternate code and text blocks, see eg in https://github.com/scikit-learn/scikit-learn/blob/master/examples/plot_compare_reduction.py

partmor commented:
> Alternative pandas solution would also be to fillna('missing') on the categorical columns (which would be equivalent to your custom transformer).

True. Should we go for this then?

> I commented just the same, so yes, I think that's a good idea :-)
I did a small GridSearchCV in my example here https://jorisvandenbossche.github.io/blog/2018/05/28/scikit-learn-columntransformer/#Using-our-pipeline-in-a-grid-search

Will go for it! Thanks for the link!!

> One other suggestion, certainly if you would make the example longer, is to use the typical sphinx gallery syntax to alternate code and text blocks, see eg in https://github.com/scikit-learn/scikit-learn/blob/master/examples/plot_compare_reduction.py

Great! I'll use this. I was starting to worry about all the # comment lines I would have to use :)


Many thanks for the feedback @jorisvandenbossche ! I'll incorporate these changes.

partmor committed:
DOC refactor with PR feedback

partmor committed:
DOC remove trailing whitespaces - pep8

jnothman commented:
Either at the top or in a comment, you should describe the nature of the features: are they numeric representations of category, or strings with how many values?

jnothman commented:
Why can't we do mode imputation in a pipeline on the categorical features after an ordinal encoding? It's a bit ugly, but it's not so bad...

jnothman commented on examples/column_transformer_mixed_types.py:
```diff
@@ -0,0 +1,94 @@
+# defined in our ``ColumnTransformer``, together with the classifier's
+# hyperparameters as part of a ``Pipeline``.
+# ``ColumnTransformer`` integrates well with the rest of scikit-learn,
+# in particular with ``GridSearchCV`.`
```
bakticks not in correct places

partmor replied:
My bad. Also, should I use sphinx syntax to reference other sklearn classes rather than simply double backticking??

jnothman commented:
This example should be referenced in place of hetero_feature_union at the
end of doc/faq.rst​

partmor committed:
DOC replace numeric for numerical in comments

partmor committed:
DOC backticks typo. add puncuations at end of comment

jorisvandenbossche commented:
> You mean?: use LabelEncoder as 1st step before one-hot encoding

No, @jnothman meant to use `CategoricalEncoder(encoding='ordinal')` (the new non-abusive version of LabelEncoder for features :-)) 

Personally I would just leave it like it is for now. The intent is that `SimpleImputer(strategy='constant')` will be available before the release, and then we can use that in this example (which results in a separate category for 'missing').

> This example should be referenced in place of hetero_feature_union at the
end of doc/faq.rst​

You can also add this example to the list of examples in the section on ColumnTransformer in compose.rst

partmor committed:
DOC add example to the list of ColumnTransformer examples in compose.rst

partmor committed:
DOC describe nature of features in comments

partmor committed:
DOC add references to Pipeline and GridSearchCV classes

partmor commented:
> Either at the top or in a comment, you should describe the nature of the features: are they numeric representations of category, or strings with how many values?

Done.

> Personally I would just leave it like it is for now. The intent is that SimpleImputer(strategy='constant') will be available before the release, and then we can use that in this example (which results in a separate category for 'missing').

I have left it as is, and in an inline comment I have mentioned the `pd.fillna` will be eventually replaced by a preprocessing step with `SimpleImputer(strategy='constant')`.

> This example should be referenced in place of hetero_feature_union at the
end of doc/faq.rst​

Done.

> You can also add this example to the list of examples in the section on ColumnTransformer in compose.rst

Done.

partmor committed:
DOC replace hetero_feature_union example in faq.rst with column_transformer_mixed_types example

amueller commented on examples/column_transformer_mixed_types.py:
```diff
@@ -0,0 +1,101 @@
+:class:`sklearn.compose.ColumnTransformer` on a dataset consisting of
+hereogeneous data types. A subset of numeric and categorical features
+are selected and :class:`sklearn.compose.ColumnTransformer` is used to
+apply two non-sequential preprocessing pipelines, one for each data
```
how about saying that different pipelines are used for different subsets of the features and the results are concatenated? I find "non-sequential" a bit odd here.

amueller commented on examples/column_transformer_mixed_types.py:
```diff
@@ -0,0 +1,101 @@
+type:
+
+* Numeric data: missing value imputation and standard scaling
+* Categorical data: missing value imputation and one-hot encoding.
```
Maybe say for categorical that missing values are replaced with a new category and for numeric data with the mean? (that's the default, right?)

amueller commented on examples/column_transformer_mixed_types.py:
```diff
@@ -0,0 +1,101 @@
+
+
+###############################################################################
+# Using our pipeline in a grid search
```
our? maybe "the"? the above?

amueller commented on examples/column_transformer_mixed_types.py:
```diff
@@ -0,0 +1,101 @@
+# We can also perform grid search on the different preprocessing steps
+# defined in our ``ColumnTransformer``, together with the classifier's
+# hyperparameters as part of a :class:`sklearn.pipeline.Pipeline`
+# ``ColumnTransformer`` integrates well with the rest of scikit-learn,
```
I think it's a bit redundant to say it integrates well. It would be quite surprising if it wouldn't.

amueller commented on examples/column_transformer_mixed_types.py:
```diff
@@ -0,0 +1,101 @@
+# We can finally fit the model using the best hyperparameters found
+# and assess model performance on holdout set
+print(("best logistic regression from grid search: %f"
+       % grid_search.best_estimator_.score(X_test, y_test)))
```
I consider this an antipattern. Use ``grid_search.score(X_test, y_test)`` unless you want to explicitly overwrite the grid-search scoring (in this case accuracy) with the estimators scoring (also accuracy). Also the comment is misaligned, the fitting happens above it. Maybe just remove the comment as the the code is pretty self-explanatory imho.

amueller commented:
looks good apart from nitpicks (and sorry about individual comments). I think the barrage of underscores in the grid-search is telling. But I'm happy we can finally actually do this!

partmor commented:
@amueller thank you for the feedback! I'll walk through the suggestions and leave it well-polished :-)

partmor committed:
DOC improve example explanation. reviewed PR feedback

ogrisel commented on examples/column_transformer_mixed_types.py:
```diff
@@ -0,0 +1,104 @@
+# - sex: categories encoded as strings {'female', 'male'}.
+# - plcass: categories encoded as ints {1, 2, 3}.
+num_feats = ['age', 'fare']
+cat_feats = ['embarked', 'sex', 'pclass']
```
Better be explicit and call those variables: `numerical_features` and `categorical_features`. Readability is extra important for examples.

partmor replied:
I agree.

ogrisel commented on examples/column_transformer_mixed_types.py:
```diff
@@ -0,0 +1,104 @@
+
+# We create the preprocessing pipelines for both numeric and categorical data.
+num_pl = make_pipeline(SimpleImputer(), StandardScaler())
+cat_pl = CategoricalEncoder('onehot-dense', handle_unknown='ignore')
```
Similar remark here; may I suggest: `numerical_transformer` and `categorical_transformer`.

partmor replied:
I agree.

ogrisel commented on examples/column_transformer_mixed_types.py:
```diff
@@ -0,0 +1,104 @@
+
+# Append classifier to preprocessing pipeline.
+# Now we have a full prediction pipeline.
+clf_pl = make_pipeline(preprocessing_pl, LogisticRegression())
```
Either `clf` or `pipeline`.

partmor replied:
I agree.

ogrisel reviewed (commented):
Looks good as well. Here are a few more comments to improve the readability even further.

ogrisel commented on examples/column_transformer_mixed_types.py:
```diff
@@ -0,0 +1,104 @@
+
+# Provisionally, use pd.fillna() to impute missing values for categorical
+# features; SimpleImputer will eventually support strategy="constant".
+data.loc[:, cat_feats] = data.loc[:, cat_feats].fillna(value='missing')
```
Do we really need the `.loc[:, columns]` pattern here? @jorisvandenbossche what is your opinion on using:

```python
data[categorical_features] = data[categorical_features].fillna(value='missing)'
```

jorisvandenbossche replied:
Yes, that would work the same
(I personally like that more (and would do it like that in my code), but it's a bit subjective)

partmor replied:
I will change it to make it more readable, it does work the same. I've had traumas with setting copy warnings from pandas, that's why I sometimes use .loc unconsciously :-)

partmor committed:
DOC improve readability of variable names

partmor committed:
resolve conflicts with master

partmor commented:
There were conflicts with a recent merge to master: I have moved the example (and its references) under the compose folder.

jorisvandenbossche commented on examples/compose/column_transformer_mixed_types.py:
```diff
@@ -0,0 +1,104 @@
+
+In this example, the numeric data is standard-scaled after
+mean-imputation, while the categorical data is one-hot
+endcoded after imputing missing values with a new category
```
typo: endcoded -> encoded

jnothman commented on examples/compose/column_transformer_mixed_types.py:
```diff
@@ -0,0 +1,104 @@
+
+# Read data from Titanic dataset.
+titanic_url = ('https://raw.githubusercontent.com/amueller/'
+               'scipy-2017-sklearn/master/notebooks/datasets/titanic3.csv')
```
maybe it's better practice to replace `master` with a fixed commit hash, i.e. `e7d90ee`

jnothman reviewed (approved):
Looks good, pending improvements to Imputer (perhaps #11211) and to ColumnTransformer (#11190)

jorisvandenbossche commented on examples/compose/column_transformer_mixed_types.py:
```diff
@@ -0,0 +1,104 @@
+In this example, the numeric data is standard-scaled after
+mean-imputation, while the categorical data is one-hot
+endcoded after imputing missing values with a new category
+(:code:`'missing'`).
```
What is the `:code:` doing? (I haven't seen this sphinx role before) 
I think it can simply be double backticks

partmor replied:
Same as double backticks, yep. Will use the latter.

jorisvandenbossche reviewed (approved):
Also minor remarks, for the rest looks good!

partmor committed:
DOC correct typos and fixed url for dataset

partmor commented:
Updated with your comments :)

jorisvandenbossche commented:
> Looks good, pending improvements to Imputer (perhaps #11211) and to ColumnTransformer (#11190)

@jnothman do you want to merge those PRs first so this example can be updated? 

I think this PR is ready to be merged, so I would do that and then those other PRs can update the example.

jnothman commented:
Thanks @partmor

jnothman commented:
Thanks @partmor

partmor commented:
Thank you all for your guidance and patience @jnothman @jorisvandenbossche @amueller @ogrisel !!!
This is my first contribution ever to a big open source project!! this only encourages me to keep going!!!
Thanks again!

partmor commented:
> Looks good, pending improvements to Imputer (perhaps #11211) and to ColumnTransformer (#11190)

I'll stay tuned to when `strategy='constant'` is implemented in `SimpleImputer`, and make a PR to include it in the example.

jnothman commented:
and congratulations Pedro. we look forward to more contributions from you
here or elsewhere in the community.

jorisvandenbossche commented:
> sorry Joris, i didn't see your comment. I don't see anything wrong with updating those PRs (though one isn't a PR yet) instead of waiting on them.

No problem, you already did exactly what I was suggesting in the meantime :-)

ogrisel commented:
Thanks @partmor!

qinhanmin2014 commented:
Question here: will it be better to put the data under scikit-learn/examples-data?
(1) It seems bit strange to fetch data from other's repo (though I respect amueller, one of our leaders)
(2) It's good for scikit-learn to have a dataset with missing data and string-encoded categorical variables (see #8888)
(3) We can also provide users with the source and some basic information of the data in the same repo.

jorisvandenbossche commented:
I think the idea is that the (future) openml loader will provide such datasets (instead of keep adding new single datasets to sklearn for each use case, although a heterogeneous dataset with missing values is of course an important use case ..)

qinhanmin2014 commented:
> I think the idea is that the (future) openml loader will provide such datasets

That's fine. Thanks a lot for the instant feedback :)

amueller commented:
Yeah it's weird to use my repo, I totally agree. Hopefully we can get rid of that soon. We could just ship the csv without a loader to use in examples (because CSV is such a common format) but that seems slightly strange?
Having an example that contains the boilerplate for separating a target from the data seems very helpful to me. Most data that I worked with required that, but none of our examples require that.
